### PR TITLE
Maintenance: replace Revision by RevisionRecord

### PR DESCRIPTION
**Pull request by Seb35**
This is a breaking change to follow Multi-Content Revisions from MediaWiki (i.e. a revision has multiple “slots”, each one is a content).

The previous Revision is replaced by MediaWiki\Revision\RevisionRecord, which is mostly similar (although not a subclass) except there are no static methods (newFromId, getTimestampFromId) and it must be specified in the method getContent the requested slot (MAIN, at least for now).

SMW has an almost-central management of revisions with RevisionGuard.

An upstream MediaWiki hook is replaced: NewRevisionFromEditComplete → RevisionFromEditComplete

A downstream Semantic MediaWiki hook has a modified signature: SMW::RevisionGuard::ChangeRevision( Title, ?RevisionRecord& )

Issue: #4944
Issue: #4941

---

**DRAFT - do not merge!**

> This first draft pull request replaced most of the deprecated class `Revision` and it seems to work with some smoke tests, but it remains details to be checked.

TODO from easiest to most difficult:
- [x] mock RevisionRecord: the two methods `getSize` and `getSha1` must be mocked else the class remains abstract, see Travis
- [ ] mock in ProtectionValidatorTest to fix the warnings?
- [ ] it seems there are issues in deleting the pages from the previous test case in `InterwikiDBIntegrationTest`, fix it
- [ ] `getContentHandler`: check/change the call to this method 
- [ ] `Revision::newFromId`: check/change the replacements - should it use RevisionGuard?
- [ ] should it be kept the `LogicException` added in RevisionGuard, or a fallback to the global `MediaWikiServices::getInstance()->getRevisionLookup()->getRevisionByTitle(…)`?
- [ ] `getTimestampFromId`: check/change the call to this method: the change in this commit is only to avoid the warning, but it should be assert if this use of RevisionStore (the only occurence in SMW) is the right way to do (see also #4714)
- [ ] use of downstream hook `SMW::RevisionGuard::ChangeRevision`: should a new hook be created or do we change signature of this existing one? - I only found SemanticApprovedRevs as public user of this hook in SMW + Gerrit repos
- [ ] keep the two old hooks? (depends on compatibility pre-1.35)

### CODE REVIEW

**Comment by Seb35** [comment]
I fixed most of the remaining errors in Travis, it remains, for 1.35, 2 failures and 2 warnings for 21782 assertions (there were yesterday 21 errors, 3 failures, 3 warnings for 21392 assertions), and for  1.32-1.34, 1 failure and 2 warnings. I did not really work to solve MW master issues even if I followed the counts (since yesterday: +38 assertions, -6 errors, +3 failures, -1 warning).

I incorporated in this draft PR the other PRs #4995 and #4996 to reduce the number of warnings and be able to see results on MW master, but they should be stripped out before the final merge given they are issues unrelated to RevisionRecord.

File: tests/phpunit/Unit/Serializers/QueryResultSerializerTest.php [hunk 1]
```diff
@@ -237,7 +237,9 @@ private function buildMockQueryResult( $setup ) {
 			] );
 
 			$printRequests[] = $printRequest;
-			$getResults[] = \SMW\DIWikiPage::newFromTitle( new \Title( NS_MAIN, $value['printRequest'] ) );
+			# TODO: Should be the following line, but the environment of the test should be adapted
+			# $getResults[] = \SMW\DIWikiPage::newFromTitle( \Title::makeTitle( NS_MAIN, $value['printRequest'] ) );
+			$getResults[] = \SMW\DIWikiPage::newFromTitle( \Title::makeTitle( NS_MAIN, '' ) );
```
Comment (Seb35) [inline tests/phpunit/Unit/Serializers/QueryResultSerializerTest.php:242]: This is an issue in the test: it is tested against an empty title. Previously it was created a Title object with `new Title()`, the constructor of Title is now private, so I used `Title::makeTitle( NS_MAIN, '' )` to continue creating it with the same behaviour, but it is a quickfix and the test should instead be adapted.

**Comment by JeroenDeDauw** [comment]
Can this be closed?

**Comment by Seb35** [comment]
This was fixed by other commits, whose #5046, #5049, #5051.

**Pull request closed without merging**